File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/xmi/XMIModelSerializer.java
package org.kevoree.modeling.api.xmi;

import org.kevoree.modeling.api.*;
import org.kevoree.modeling.api.meta.MetaAttribute;
import org.kevoree.modeling.api.meta.MetaReference;
import org.kevoree.modeling.api.util.CallBackChain;
import org.kevoree.modeling.api.util.Helper;

/*
* Author : Gregory Nain
* Date : 02/09/13
*/

public class XMIModelSerializer {

    public static void save(KObject model, final ThrowableCallback<String> callback) {
        final SerializationContext context = new SerializationContext();
        context.model = model;
        context.finishCallback = callback;
        context.attributesVisitor = new ModelAttributeVisitor() {
            @Override
            public void visit(MetaAttribute metaAttribute, Object value) {
                if (value != null) {
                    if (context.ignoreGeneratedID && metaAttribute.metaName().equals("generated_KMF_ID")) {
                        return;
                    }
                    context.printer.append(" " + metaAttribute.metaName() + "=\"");
                    XMIModelSerializer.escapeXml(context.printer, value.toString());
                    context.printer.append("\"");
                }
            }
        };

        context.printer = new StringBuilder();
        //First Pass for building address table
        context.addressTable.put(model.uuid(), "/");


        KTask addressCreationTask = context.model.taskVisit(new ModelVisitor() {
            @Override
            public VisitResult visit(KObject elem) {
                String parentXmiAddress = context.addressTable.get(elem.parentUuid());
                String key = parentXmiAddress + "/@" + elem.referenceInParent().metaName();
                Integer i = context.elementsCount.get(key);
                if (i == null) {
                    i = 0;
                    context.elementsCount.put(key, i);
                }
                context.addressTable.put(elem.uuid(), parentXmiAddress + "/@" + elem.referenceInParent().metaName() + "." + i);
                context.elementsCount.put(parentXmiAddress + "/@" + elem.referenceInParent().metaName(), i + 1);
                String pack = elem.metaClass().metaName().substring(0, elem.metaClass().metaName().lastIndexOf('.'));
                if (!context.packageList.contains(pack)) {
                    context.packageList.add(pack);
                }
                return VisitResult.CONTINUE;
            }
        }, VisitRequest.CONTAINED);

        KTask serializationTask = context.model.universe().model().task();
        serializationTask.wait(addressCreationTask);
        serializationTask.setJob(new KJob() {
            @Override
            public void run(KCurrentTask currentTask) {
                context.printer.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
                context.printer.append("<" + XMIModelSerializer.formatMetaClassName(context.model.metaClass().metaName()).replace(".", "_"));
                context.printer.append(" xmlns:xsi=\"http://wwww.w3.org/2001/XMLSchema-instance\"");
                context.printer.append(" xmi:version=\"2.0\"");
                context.printer.append(" xmlns:xmi=\"http://www.omg.org/XMI\"");

                int index = 0;
                while (index < context.packageList.size()) {
                    context.printer.append(" xmlns:" + context.packageList.get(index).replace(".", "_") + "=\"http://" + context.packageList.get(index) + "\"");
                    index++;
                }
                context.model.visitAttributes(context.attributesVisitor);

                KTask nonContainedRefsTasks = context.model.universe().model().task();
                for (int i = 0; i < context.model.metaClass().metaReferences().length; i++) {
                    if (!context.model.metaClass().metaReferences()[i].contained()) {
                        nonContainedRefsTasks.wait(nonContainedReferenceTaskMaker(context.model.metaClass().metaReferences()[i], context, context.model));
                    }
                }
                nonContainedRefsTasks.setJob(new KJob() {
                    @Override
                    public void run(KCurrentTask currentTask) {
                        context.printer.append(">\n");

                        KTask containedRefsTasks = context.model.universe().model().task();
                        for (int i = 0; i < context.model.metaClass().metaReferences().length; i++) {
                            if (context.model.metaClass().metaReferences()[i].contained()) {
                                containedRefsTasks.wait(containedReferenceTaskMaker(context.model.metaClass().metaReferences()[i], context, context.model));
                            }
                        }
                        containedRefsTasks.setJob(new KJob() {
                            @Override
                            public void run(KCurrentTask currentTask) {
                                context.printer.append("</" + XMIModelSerializer.formatMetaClassName(context.model.metaClass().metaName()).replace(".", "_") + ">\n");
                                context.finishCallback.on(context.printer.toString(), null);
                            }
                        });
                        containedRefsTasks.ready();
                    }
                });
                nonContainedRefsTasks.ready();
            }
        });
        serializationTask.ready();
    }


    public static void escapeXml(StringBuilder ostream, String chain) {
        if (chain == null) {
            return;
        }
        int i = 0;
        int max = chain.length();
        while (i < max) {
            char c = chain.charAt(i);
            if (c == '"') {
                ostream.append("&quot;");
            } else if (c == '&') {
                ostream.append("&amp;");
            } else if (c == '\'') {
                ostream.append("&apos;");
            } else if (c == '<') {
                ostream.append("&lt;");
            } else if (c == '>') {
                ostream.append("&gt;");
            } else {
                ostream.append(c);
            }
            i = i + 1;
        }
    }

    public static String formatMetaClassName(String metaClassName) {
        int lastPoint = metaClassName.lastIndexOf('.');
        String pack = metaClassName.substring(0, lastPoint);
        String cls = metaClassName.substring(lastPoint + 1);
        return pack + ":" + cls;
    }

    private static KTask nonContainedReferenceTaskMaker(final MetaReference ref, SerializationContext p_context, KObject p_currentElement) {
        KTask allTask = p_currentElement.taskAll(ref);

        KTask thisTask = p_context.model.universe().model().task();
        thisTask.wait(allTask);

        thisTask.setJob(new KJob() {
            @Override
            public void run(KCurrentTask currentTask) {
                try {
                    KObject[] objects = ((KObject[]) currentTask.results().get(allTask));
                    for (int i = 0; i < objects.length; i++) {
                        String adjustedAddress = p_context.addressTable.get(objects[i].uuid());
                        p_context.printer.append(" " + ref.metaName() + "=\"" + adjustedAddress + "\"");
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
        thisTask.ready();
        return thisTask;
    }

    private static KTask containedReferenceTaskMaker(final MetaReference ref, SerializationContext context, KObject currentElement) {
        KTask allTask = currentElement.taskAll(ref);

        KTask thisTask = context.model.universe().model().task();
        thisTask.wait(allTask);
        thisTask.setJob(new KJob() {
            @Override
            public void run(KCurrentTask currentTask) {
                try {
                    if (currentTask.results().get(allTask) != null) {
                        KObject[] objs =  ((KObject[]) currentTask.results().get(allTask));
                        for (int i = 0; i < objs.length; i++) {
                            final KObject elem = objs[i];
                            context.printer.append("<");
                            context.printer.append(ref.metaName());
                            context.printer.append(" xsi:type=\"" + XMIModelSerializer.formatMetaClassName(elem.metaClass().metaName()) + "\"");
                            elem.visitAttributes(context.attributesVisitor);

                            KTask nonContainedRefsTasks = context.model.universe().model().task();
                            for (int j = 0; j < elem.metaClass().metaReferences().length; j++) {
                                if (!elem.metaClass().metaReferences()[i].contained()) {
                                    nonContainedRefsTasks.wait(nonContainedReferenceTaskMaker(elem.metaClass().metaReferences()[i], context, elem));
                                }
                            }
                            nonContainedRefsTasks.setJob(new KJob() {
                                @Override
                                public void run(KCurrentTask currentTask) {
                                    context.printer.append(">\n");


                                    KTask containedRefsTasks = context.model.universe().model().task();
                                    for (int i = 0; i < elem.metaClass().metaReferences().length; i++) {
                                        if (elem.metaClass().metaReferences()[i].contained()) {
                                            containedRefsTasks.wait(containedReferenceTaskMaker(elem.metaClass().metaReferences()[i], context, elem));
                                        }
                                    }
                                    containedRefsTasks.setJob(new KJob() {
                                        @Override
                                        public void run(KCurrentTask currentTask) {
                                            context.printer.append("</");
                                            context.printer.append(ref.metaName());
                                            context.printer.append('>');
                                            context.printer.append("\n");
                                        }
                                    });
                                    containedRefsTasks.ready();
                                }
                            });
                            nonContainedRefsTasks.ready();
                        }
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
        thisTask.ready();
        return thisTask;
    }
}

File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/xmi/XMIModelSerializer.kt
package org.kevoree.modeling.api.xmi

import org.kevoree.modeling.api.util.ModelVisitor
import org.kevoree.modeling.api.KMFContainer
import org.kevoree.modeling.api.util.ModelAttributeVisitor
import org.kevoree.modeling.api.util.AttConverter

/*
* Author : Gregory Nain
* Date : 02/09/13
*/
class ReferencesVisitor(val ostream: java.io.PrintStream, val addressTable: java.util.HashMap<KMFContainer, String>, val elementsCount: java.util.HashMap<String, Int>, val resourceSet: ResourceSet?) : ModelVisitor() {

    var value: String? = null

    override fun endVisitRef(refName: String) {
        if (value != null) {
            ostream.print(" " + refName + "=\"" + value.toString() + "\"")
            value = null
        }
    }

    public override fun visit(elem: KMFContainer, refNameInParent: String, parent: KMFContainer) {
        var adjustedAddress = resourceSet?.objToAddr(elem)
        if (adjustedAddress == null) {
            adjustedAddress = addressTable.get(elem)
        }
        if (value == null) {
            value = adjustedAddress
        } else {
            value = value!! + " "
            value = value!! + adjustedAddress.toString()
        }
    }
}

class AttributesVisitor(val ostream: java.io.PrintStream, val ignoreGeneratedID: Boolean) : ModelAttributeVisitor {
    public override fun visit(value: Any?, name: String, parent: KMFContainer) {
        if (value != null) {
            if (ignoreGeneratedID && name == "generated_KMF_ID") {
                return
            }
            if (value is String && value == "") {
                return
            }
            ostream.print(" " + name + "=\"")
            if (value is java.util.Date) {
                escapeXml(ostream, "" + value.getTime())
            } else {
                escapeXml(ostream, AttConverter.convFlatAtt(value))
            }
            ostream.print("\"")
        }
    }

    private fun escapeXml(ostream: java.io.PrintStream, chain: String?) {
        if (chain == null) {
            return
        }
        var i = 0
        var max = chain.length
        while (i < max) {
            var c = chain.charAt(i)
            if (c == '"') {
                ostream.print("&quot;")
            } else if (c == '&') {
                ostream.print("&amp;")
            } else if (c == '\'') {
                ostream.print("&apos;")
            } else if (c == '<') {
                ostream.print("&lt;")
            } else if (c == '>') {
                ostream.print("&gt;")
            } else {
                ostream.print(c)
            }
            i = i + 1
        }
    }
}



class ModelSerializationVisitor(val ostream: java.io.PrintStream, val addressTable: java.util.HashMap<org.kevoree.modeling.api.KMFContainer, String>, val elementsCount: java.util.HashMap<String, Int>, val resourceSet: ResourceSet?, ignoreGeneratedID: Boolean) : ModelVisitor() {
    val attributeVisitor = AttributesVisitor(ostream, ignoreGeneratedID)
    val referenceVisitor = ReferencesVisitor(ostream, addressTable, elementsCount, resourceSet)

    public override fun visit(elem: KMFContainer, refNameInParent: String, parent: KMFContainer) {
        ostream.print('<')
        ostream.print(refNameInParent)
        ostream.print(" xsi:type=\"" + formatMetaClassName(elem.metaClassName()) + "\"")
        elem.visitAttributes(attributeVisitor)
        elem.visit(referenceVisitor, false, false, true)
        ostream.println('>')
        elem.visit(this, false, true, false)
        ostream.print("</")
        ostream.print(refNameInParent)
        ostream.print('>')
        ostream.println()
    }

    private fun formatMetaClassName(metaClassName: String): String {
        val lastPoint = metaClassName.lastIndexOf('.')
        val pack = metaClassName.substring(0, lastPoint)
        val cls = metaClassName.substring(lastPoint + 1)
        return pack + ":" + cls
    }
}


class ModelAddressVisitor(val addressTable: java.util.HashMap<KMFContainer, String>, val elementsCount: java.util.HashMap<String, Int>, val packageList: java.util.ArrayList<String>) : ModelVisitor() {
    public override fun visit(elem: KMFContainer, refNameInParent: String, parent: KMFContainer) {

        val parentXmiAddress = addressTable.get(parent)!!
        val i = elementsCount.get(parentXmiAddress + "/@" + refNameInParent) ?: 0
        addressTable.put(elem, parentXmiAddress + "/@" + refNameInParent + "." + i)
        elementsCount.put(parentXmiAddress + "/@" + refNameInParent, i + 1)
        val pack = elem.metaClassName().substring(0, elem.metaClassName().lastIndexOf('.'))
        if (!packageList.contains(pack))
            packageList.add(pack)
    }
}




public class XMIModelSerializer() : org.kevoree.modeling.api.ModelSerializer {

    public var resourceSet: ResourceSet? = null

    public var ignoreGeneratedID: Boolean = false

    override fun serialize(oMS: org.kevoree.modeling.api.KMFContainer): String? {
        val oo = java.io.ByteArrayOutputStream()
        serializeToStream(oMS, oo)
        oo.flush()
        return oo.toString()
    }
    override fun serializeToStream(oMS: org.kevoree.modeling.api.KMFContainer, ostream: java.io.OutputStream) {

        val wt = java.io.PrintStream(java.io.BufferedOutputStream(ostream), false)

        //First Pass for building address table
        val addressTable: java.util.HashMap<org.kevoree.modeling.api.KMFContainer, String> = java.util.HashMap<org.kevoree.modeling.api.KMFContainer, String>()
        val packageList: java.util.ArrayList<String> = java.util.ArrayList<String>()
        addressTable.put(oMS, "/")
        val elementsCount: java.util.HashMap<String, Int> = java.util.HashMap<String, Int>()
        var addressBuilderVisitor = ModelAddressVisitor(addressTable, elementsCount, packageList)
        oMS.visit(addressBuilderVisitor, true, true, false)

        var masterVisitor = ModelSerializationVisitor(wt, addressTable, elementsCount, resourceSet, ignoreGeneratedID)

        wt.println("<?xml version=\"1.0\" encoding=\"UTF-8\"?>")

        wt.print("<" + formatMetaClassName(oMS.metaClassName()).replace(".", "_"))
        wt.print(" xmlns:xsi=\"http://wwww.w3.org/2001/XMLSchema-instance\"")
        wt.print(" xmi:version=\"2.0\"")
        wt.print(" xmlns:xmi=\"http://www.omg.org/XMI\"")

        var index = 0;
        while (index < packageList.size) {
            wt.print(" xmlns:" + packageList.get(index).replace(".", "_") + "=\"http://" + packageList.get(index) + "\"")
            index++
        }

        oMS.visitAttributes(AttributesVisitor(wt, ignoreGeneratedID))
        oMS.visit(ReferencesVisitor(wt, addressTable, elementsCount, resourceSet), false, false, true)
        wt.println(">")

        oMS.visit(masterVisitor, false, true, false)

        wt.println("</" + formatMetaClassName(oMS.metaClassName()).replace(".", "_") + ">")

        wt.flush()
    }

    private fun formatMetaClassName(metaClassName: String): String {
        val lastPoint = metaClassName.lastIndexOf('.')
        val pack = metaClassName.substring(0, lastPoint)
        val cls = metaClassName.substring(lastPoint + 1)
        return pack + ":" + cls
    }


}